Included application upgrade funclib utils file and its associated litmusbook .

vibhor995 opened the pull request:
Signed-off-by: vibhor995 <vibhor.kumar@mayadata.io>

<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:

- Included functional library for the application upgrade test job.
-  Also included its associated test-jobs.
    (Used jenkins as a test application )




**Special notes for your reviewer**:
1) This PR also includes changes in ansible-runner dockerfile (included funclib path ).
2) This test is using jenkins images "jenkins/jenkins:2.124".

@gprasath @dargasudarshan @ksatchit @nsathyaseelan

vibhor995 commented:
**output**
```No config file found; using defaults
[DEPRECATION WARNING]: Specifying include variables at the top-level of the 
task is deprecated. Please see: 
https://docs.ansible.com/ansible/playbooks_roles.html#task-include-files-and-
encouraging-reuse   for currently supported syntax regarding included files and
 variables. This feature will be removed in version 2.12. Deprecation warnings 
can be disabled by setting deprecation_warnings=False in ansible.cfg.

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Record the funcutil  path] ***********************************************
ok: [127.0.0.1] => {"ansible_facts": {"funclib_path": "/funclib/kubectl/app_upgrade.yml"}, "changed": false}

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "7410303c05bc0b807d5804bda5a459250c5a79b3", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "bc2b5f20c8828a5ed15a3772d5eb0cfc", "mode": "0644", "owner": "root", "size": 421, "src": "/root/.ansible/tmp/ansible-tmp-1538638912.62-183811486328318/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.349244", "end": "2018-10-04 07:41:54.756361", "failed_when_result": false, "rc": 0, "start": "2018-10-04 07:41:53.407117", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"jenkins-upgrade-test\" created", "stdout_lines": ["litmusresult \"jenkins-upgrade-test\" created"]}

TASK [Display the app information passed via the test job] *********************
ok: [127.0.0.1] => {
    "msg": [
        "The application info is as follows:", 
        "Namespace           : jenkins", 
        "Label               : app=jenkins-app"
    ]
}

TASK [Verify that the AUT (Application Under Test) is running] *****************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n jenkins -l app=jenkins-app --no-headers -o custom-columns=:status.phase", "delta": "0:00:01.004800", "end": "2018-10-04 07:41:56.016134", "rc": 0, "start": "2018-10-04 07:41:55.011334", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [include] *****************************************************************
included: /funclib/kubectl/app_upgrade.yml for 127.0.0.1

TASK [Obtaining the deployment name.] ******************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployment -n jenkins --no-headers -l app=jenkins-app -o custom-columns=:metadata.name", "delta": "0:00:00.982390", "end": "2018-10-04 07:41:57.251495", "rc": 0, "start": "2018-10-04 07:41:56.269105", "stderr": "", "stderr_lines": [], "stdout": "jenkins", "stdout_lines": ["jenkins"]}

TASK [Recording the application pod name.] *************************************
ok: [127.0.0.1] => {"ansible_facts": {"app_name": "jenkins"}, "changed": false}

TASK [Updating image of application to a required version.] ********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl set image deployment/jenkins jenkins=jenkins/jenkins:2.124 -n jenkins", "delta": "0:00:00.970522", "end": "2018-10-04 07:41:58.465330", "failed_when_result": false, "rc": 0, "start": "2018-10-04 07:41:57.494808", "stderr": "", "stderr_lines": [], "stdout": "deployment \"jenkins\" image updated", "stdout_lines": ["deployment \"jenkins\" image updated"]}

TASK [Check if  the application pod is running after upgrade.] *****************
FAILED - RETRYING: Check if  the application pod is running after upgrade. (10 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pod -n jenkins --no-headers -l app=jenkins-app -o jsonpath='{.items[0].status.phase}'", "delta": "0:00:00.973915", "end": "2018-10-04 07:42:11.014023", "rc": 0, "start": "2018-10-04 07:42:10.040108", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking status of application pod] **************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod -n jenkins -l app=jenkins-app -o jsonpath='{.items[0].status.phase}'", "delta": "0:00:01.026993", "end": "2018-10-04 07:42:12.253144", "rc": 0, "start": "2018-10-04 07:42:11.226151", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking the image of application pod post upgrade.] *********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployment jenkins -n jenkins -o jsonpath=\"{..image}\"", "delta": "0:00:01.001609", "end": "2018-10-04 07:42:13.485558", "failed_when_result": false, "rc": 0, "start": "2018-10-04 07:42:12.483949", "stderr": "", "stderr_lines": [], "stdout": "jenkins/jenkins:2.124", "stdout_lines": ["jenkins/jenkins:2.124"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "47389d5ce23f69de7c7861ce15068d8154bea06e", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "e27446b0c95f4993b46a996b66c627fe", "mode": "0644", "owner": "root", "size": 419, "src": "/root/.ansible/tmp/ansible-tmp-1538638933.62-19398180734948/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.043809", "end": "2018-10-04 07:42:16.676172", "failed_when_result": false, "rc": 0, "start": "2018-10-04 07:42:15.632363", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"jenkins-upgrade-test\" configured", "stdout_lines": ["litmusresult \"jenkins-upgrade-test\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=16   changed=10   unreachable=0    failed=0

vibhor995 commented:
**litmus CR output**

```
Name:         jenkins-upgrade-test
Namespace:
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"litmus.io/v1alpha1","kind":"LitmusResult","metadata":{"annotations":{},"name":"jenkins-upgrade-test","namespace":""},"spec":{"testMetada...
API Version:  litmus.io/v1alpha1
Kind:         LitmusResult
Metadata:
  Cluster Name:
  Creation Timestamp:  2018-10-04T07:41:54Z
  Generation:          0
  Resource Version:    6142227
  Self Link:           /apis/litmus.io/v1alpha1/jenkins-upgrade-test
  UID:                 f7031100-c7a8-11e8-8143-42010a800193
Spec:
  Test Metadata:
    App:        <nil>
    Chaostype:  <nil>
  Test Status:
    Phase:   completed
    Result:  Pass
Events:      <none>

nsathyaseelan commented on apps/jenkins/upgrade/run_litmus_test.yml:
```diff
@@ -0,0 +1,45 @@
+
+            # Either deployment or statefulset
+          - name: DEPLOYMENT_TYPE
+            value: deployment
```
@vibhor995 use DEPLOY_TYPE type instead of DEPLOYMENT_TYPE

vibhor995 replied:
@nsathyaseelan done

nsathyaseelan commented on apps/jenkins/upgrade/test_vars.yml:
```diff
@@ -0,0 +1,7 @@
+app_namespace: "{{ lookup('env','APP_NAMESPACE') }}"
+label: "{{ lookup('env','APP_LABEL') }}"
+funcutil: kubectl/app_upgrade.yml
+deploy_type: "{{ lookup('env','DEPLOYMENT_TYPE') }}"
```
@vibhor995 can you change DEPLOYMENT_TYPE into DEPLOY_TYPE here also

nsathyaseelan commented on tools/ansible-runner/Dockerfile:
```diff
@@ -24,6 +24,8 @@ ADD ./apps/ ./
 COPY ./chaoslib ./chaoslib/
 
+COPY ./funclib ./funclib
+
```
@vibhor995 get the latest changes for the docker file and update the changes

nsathyaseelan commented on apps/jenkins/upgrade/run_litmus_test.yml:
```diff
@@ -0,0 +1,45 @@
+
+            # Specified image name for application
+          - name: IMAGE_NAME
+            value:   
```
@vibhor995 Can we specify default image for app upgrade?

vibhor995 replied:
@nsathyaseelan we can only specify the default image name when we are sure that our existing app is going to deployed with certain image .

dargasudarshan commented on apps/jenkins/upgrade/run_litmus_test.yml:
```diff
@@ -0,0 +1,45 @@
+
+            # Application namespace
+          - name: APP_NAMESPACE
+            value: jenkins
```
@vibhor995 Can you update the app namespace with the default value?

dargasudarshan commented on funclib/kubectl/app_upgrade.yml:
```diff
@@ -0,0 +1,34 @@
+  failed_when: "'image updated' not in result.stdout"
+
+- name: Check if  the application pod is running after upgrade.
+  shell: kubectl get pod -n {{ app_name  }} --no-headers -l {{ app_label }} -o jsonpath='{.items[0].status.phase}'
```
@vibhor995 Here we should pass {{ app_ns }}, not {{ app_name }}

dargasudarshan commented:
@vibhor995 Can you resolve the merge conflicts

vibhor995 commented:
**output when image is different from running app**

```

No config file found; using defaults
[DEPRECATION WARNING]: Specifying include variables at the top-level of the 
task is deprecated. Please see: 
https://docs.ansible.com/ansible/playbooks_roles.html#task-include-files-and-
encouraging-reuse   for currently supported syntax regarding included files and
 variables. This feature will be removed in version 2.12. Deprecation warnings 
can be disabled by setting deprecation_warnings=False in ansible.cfg.

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Record the funcutil  path] ***********************************************
ok: [127.0.0.1] => {"ansible_facts": {"funclib_path": "/funclib/kubectl/app_upgrade.yml"}, "changed": false}

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "5c31736d47bcb1f305a5a88372a04ccd85856955", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "e332cfb3d4101aa0ae7f634386dd6cef", "mode": "0644", "owner": "root", "size": 428, "src": "/root/.ansible/tmp/ansible-tmp-1539166703.91-79382159538422/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.315902", "end": "2018-10-10 10:18:25.993887", "failed_when_result": false, "rc": 0, "start": "2018-10-10 10:18:24.677985", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/jenkins-upgrade-test configured", "stdout_lines": ["litmusresult.litmus.io/jenkins-upgrade-test configured"]}

TASK [Display the app information passed via the test job] *********************
ok: [127.0.0.1] => {
    "msg": [
        "The application info is as follows:", 
        "Namespace           : app-jenkins-ns", 
        "Label               : app=jenkins-app"
    ]
}

TASK [Verify that the AUT (Application Under Test) is running] *****************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n app-jenkins-ns -l app=jenkins-app --no-headers -o custom-columns=:status.phase", "delta": "0:00:00.838983", "end": "2018-10-10 10:18:27.106176", "rc": 0, "start": "2018-10-10 10:18:26.267193", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [include] *****************************************************************
included: /funclib/kubectl/app_upgrade.yml for 127.0.0.1

TASK [Obtaining the deployment name.] ******************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployment -n app-jenkins-ns --no-headers -l app=jenkins-app -o custom-columns=:metadata.name", "delta": "0:00:00.865591", "end": "2018-10-10 10:18:28.237686", "rc": 0, "start": "2018-10-10 10:18:27.372095", "stderr": "", "stderr_lines": [], "stdout": "jenkins", "stdout_lines": ["jenkins"]}

TASK [Recording the application pod name.] *************************************
ok: [127.0.0.1] => {"ansible_facts": {"app_name": "jenkins"}, "changed": false}

TASK [Get the image of application pre-upgrade.] *******************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployment jenkins -n app-jenkins-ns -o jsonpath=\"{..image}\"", "delta": "0:00:00.847650", "end": "2018-10-10 10:18:29.340764", "rc": 0, "start": "2018-10-10 10:18:28.493114", "stderr": "", "stderr_lines": [], "stdout": "jenkins/jenkins:2.125", "stdout_lines": ["jenkins/jenkins:2.125"]}

TASK [Record the image name of existing running application.] ******************
ok: [127.0.0.1] => {"ansible_facts": {"image": "jenkins/jenkins:2.125"}, "changed": false}

TASK [Updating image of application to a required version.] ********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl set image deployment/jenkins jenkins=jenkins/jenkins:lts -n app-jenkins-ns", "delta": "0:00:00.853037", "end": "2018-10-10 10:18:30.439224", "failed_when_result": false, "rc": 0, "start": "2018-10-10 10:18:29.586187", "stderr": "", "stderr_lines": [], "stdout": "deployment.extensions/jenkins image updated", "stdout_lines": ["deployment.extensions/jenkins image updated"]}

TASK [Check if  the application pod is running after upgrade.] *****************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod -n app-jenkins-ns --no-headers -l app=jenkins-app -o jsonpath='{.items[0].status.phase}'", "delta": "0:00:00.838369", "end": "2018-10-10 10:18:31.485910", "rc": 0, "start": "2018-10-10 10:18:30.647541", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [debug] *******************************************************************
skipping: [127.0.0.1] => {}

TASK [Checking status of application pod] **************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod -n app-jenkins-ns -l app=jenkins-app -o jsonpath='{.items[0].status.phase}'", "delta": "0:00:00.894369", "end": "2018-10-10 10:18:32.610277", "rc": 0, "start": "2018-10-10 10:18:31.715908", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking the image of application pod post upgrade.] *********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployment jenkins -n app-jenkins-ns -o jsonpath=\"{..image}\"", "delta": "0:00:00.846965", "end": "2018-10-10 10:18:33.641152", "failed_when_result": false, "rc": 0, "start": "2018-10-10 10:18:32.794187", "stderr": "", "stderr_lines": [], "stdout": "jenkins/jenkins:lts", "stdout_lines": ["jenkins/jenkins:lts"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "0c2afb16f965e00048bd773cf6f50bdc4a852938", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "86b85dec5b46ac06389c20f558031d13", "mode": "0644", "owner": "root", "size": 426, "src": "/root/.ansible/tmp/ansible-tmp-1539166713.79-109158271424535/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.027238", "end": "2018-10-10 10:18:36.702493", "failed_when_result": false, "rc": 0, "start": "2018-10-10 10:18:35.675255", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/jenkins-upgrade-test configured", "stdout_lines": ["litmusresult.litmus.io/jenkins-upgrade-test configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=18   changed=11   unreachable=0    failed=0
